From: Hailey Mothershead <hailmo@amazon.com>
Date: Thu, 29 Jun 2023 19:50:01 +0000
Subject: crpyto: testmgr.c - set fips_enabled for sha3-256, sha3-512, and
 their HMAC variants

As of 'crypto: jitter - replace LFSR with SHA3-256' this kernel does not boot in
fips mode. This is because the ecdh-nist-p256-generic selftest fails, unable to
find jitterentropy_rng. jitter is not loaded because it, during its
initialization, is not able to find sha3-256-generic. The kernel is unable to
load sha3-256 because it has not yet had a self-test run on it. Marking it as
allowed here allows the self-test to have run in time for jitter to find it on
initalization.

'Enable-Algorithims-for-Amazon-Linux-6.1.y' disallows these
algorithims downstream. It is not clear why it choses to do so other
than the algorithims being "not supported". This patch was ported from
AL2 for AL2023 - its not clear why they were considered unsupported
then. FIPs 140-3 dissalows the 224 and 384 versions. So undo this patch
for sha2-256 and sha3-512 - including the HMAC disabling.

Signed-off-by: Hailey Mothershead hailmo@amazon.com
---
 crypto/testmgr.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/crypto/testmgr.c b/crypto/testmgr.c
--- a/crypto/testmgr.c
+++ b/crypto/testmgr.c
@@ -5240,6 +5240,7 @@ static const struct alg_test_desc alg_test_descs[] = {
 	}, {
 		.alg = "hmac(sha3-256)",
 		.test = alg_test_hash,
+		.fips_allowed = 1,
 		.suite = {
 			.hash = __VECS(hmac_sha3_256_tv_template)
 		}
@@ -5252,6 +5253,7 @@ static const struct alg_test_desc alg_test_descs[] = {
 	}, {
 		.alg = "hmac(sha3-512)",
 		.test = alg_test_hash,
+		.fips_allowed = 1,
 		.suite = {
 			.hash = __VECS(hmac_sha3_512_tv_template)
 		}
@@ -5561,6 +5563,7 @@ static const struct alg_test_desc alg_test_descs[] = {
 	}, {
 		.alg = "sha3-256",
 		.test = alg_test_hash,
+		.fips_allowed = 1,
 		.suite = {
 			.hash = __VECS(sha3_256_tv_template)
 		}
@@ -5573,6 +5576,7 @@ static const struct alg_test_desc alg_test_descs[] = {
 	}, {
 		.alg = "sha3-512",
 		.test = alg_test_hash,
+		.fips_allowed = 1,
 		.suite = {
 			.hash = __VECS(sha3_512_tv_template)
 		}

